### PR TITLE
Fix memory leak issues in Calo Layer 1 UCT Infrastructure

### PR DESCRIPTION
**Pull request by aloeliger**
### PR description:

This PR should handle some of the worst offending pointer issues and memory leak problems in https://github.com/cms-sw/cmssw/issues/46526. This swaps a lot of Calo Layer 1 Infrastructure pointers over to `std::shared_ptr`'s to try and avoid memory leak issues. Summary card regions are now stored as the objects themselves, instead of pointers to them (That decision may not be directly necessary? It makes sense to me, and there is a spot where pointers to these things are used internally, and `nullptr`s are possible but checked against)

This is just a first pass, but a larger fix of all of this infrastructure with smart pointers should be considered.

@mmusich @makortel 

### PR validation:

Code compiles, and has had formatting applied

To be updated with further tests. PR still in testing.

### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

This PR is not a backport, but may need backporting to releases responsible for current prompt reco processing for patches.

### CODE REVIEW

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by aloeliger** [comment]
Weird. Didn't see the set but unused warnings when I compiled it. Let me just double check and I'll pull those out tomorrow morning.

**Comment by mmusich** [comment]
>  Didn't see the set but unused warnings when I compiled it.

Normal compilation won't show the problem. 
In order to reproduce you need to run `clang` (using the command `scram build -k -j 32 COMPILER='llvm compile'`).

**Comment by aloeliger** [comment]
I'm doing some reconsidering on all the stuff I'm ripping out, because there's a large amount of code that was dedicated to (ultimately unused) calculations of Calo Layer 1 variables that were leaking memory. I don't know whether these variables may find a use in the future, or at the very least, we dont want to lose any knowledge related to them, so I may add these back, but in a memory safe way.

**Comment by aloeliger** [comment]
I think this should be a little better.

**Comment by aloeliger** [comment]
please test

**Comment by mandrenguyen** [comment]
ignore tests-rejected with ib-failure
See https://github.com/cms-sw/cmssw/issues/46555

**Comment by mmusich** [comment]
@aloeliger what's the status of this PR?

**Comment by aloeliger** [comment]
@mmusich it "should" be fine because nothing really changed all that much, just went from pointers to smart pointers, so I could sign now with relative confidence, but I'd appreciate the chance to test it.

**Comment by mmusich** [comment]
urgent
   * there are more paused jobs at Tier0 compatible with the issue described at https://github.com/cms-sw/cmssw/issues/46526

**Comment by makortel** [comment]
Is this code run in any of the runTheMatrix workflows?

File: L1Trigger/L1TCaloLayer1/test/testUCTLayer1.cpp [hunk 1]
```diff
@@ -57,7 +57,7 @@ void print(UCTLayer1& uct) {
         if (regions[rgn]->et() > 0) {
           int hitEta = regions[rgn]->hitCaloEta();
           int hitPhi = regions[rgn]->hitCaloPhi();
-          vector<UCTTower*> towers = regions[rgn]->getTowers();
+          std::vector<std::shared_ptr<UCTTower>> towers = regions[rgn]->getTowers();
```
Comment (Dr15Jones) [inline L1Trigger/L1TCaloLayer1/test/testUCTLayer1.cpp:60]:
```suggestion
          auto const& towers = regions[rgn]->getTowers();
```
will be easier to maintain and will be more efficient as it avoids a copy.

File: L1Trigger/L1TCaloLayer1/test/testUCTLayer1HF.cpp [hunk 2]
```diff
@@ -56,7 +56,7 @@ void print(UCTLayer1& uct) {
     for (uint32_t crd = 0; crd < cards.size(); crd++) {
       vector<UCTRegion*> regions = cards[crd]->getRegions();
       for (uint32_t rgn = 0; rgn < regions.size(); rgn++) {
-        vector<UCTTower*> towers = regions[rgn]->getTowers();
+        vector<std::shared_ptr<UCTTower>> towers = regions[rgn]->getTowers();
```
Comment (Dr15Jones) [inline L1Trigger/L1TCaloLayer1/test/testUCTLayer1HF.cpp:59]:
```suggestion
        auto const& towers = regions[rgn]->getTowers();
```

**Comment by makortel** [comment]
In any case comparing the logs of 12834.0 step3 the allocated memory used at the process exit time decreases from [1681833112](https://cmssdt.cern.ch/SDT/jenkins-artifacts/ib-baseline-tests/CMSSW_14_2_X_2024-10-30-1100/el8_amd64_gcc12/-GenuineIntel/matrix-results/12834.0_TTbar_14TeV+2024/step3_TTbar_14TeV+2024.log) bytes to [1681269224](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-4cff68/42471/runTheMatrix-results/12834.0_TTbar_14TeV+2024/step3_TTbar_14TeV+2024.log), i.e. by 551 kB, which is close to the expected reduction of 497 for 10 events. On the other hand, the relative reduction is 0.03 %, and I don't know if the MaxMemoryPreload report is that precise.

**Comment by mmusich** [comment]
FWIW I've run 100k events out of the "tarball" reported from Tier0 (see [description](https://github.com/cms-sw/cmssw/issues/46526#issue-2618104325)) and compared the RSS with / without this PR:

image

things seem to go in the right direction.

**Comment by Dr15Jones** [comment]
So I also did a test using a trivial fast workflow (Neutrino beam as start) and used the MaxAllocMonitor before and after using this fix

image

**Comment by mmusich** [comment]
@aloeliger, please address the review comments https://github.com/cms-sw/cmssw/pull/46543#discussion_r1824113109 and https://github.com/cms-sw/cmssw/pull/46543#discussion_r1824114005 at earliest convenience.

>  but I'd appreciate the chance to test it.

can you clarify what's left to test?
Also we need urgently a backport to CMSSW_14_1_X to build a patch for the incoming PbPb operations (as early as next Monday).

**Comment by mandrenguyen** [comment]
Indeed, I would like to build a new 14_1_X tonight.  If we can get the backport of this today, would be great.

**Comment by aloeliger** [comment]
My apologies for all the delays, I'm in the middle of moving from CERN back to the US. I wanted to just make sure CICADA emulation was the same before and after this PR but I think given the urgency I can forgo that. I'll make the requested changes when I'm back in front of computer, and make the backport. ~2 hours maybe?

File: L1Trigger/L1TCaloLayer1/test/testUCTLayer1.cpp (hunk 1 above)
Comment (aloeliger) [reply L1Trigger/L1TCaloLayer1/test/testUCTLayer1.cpp]: Applied in most recent commit

File: L1Trigger/L1TCaloLayer1/test/testUCTLayer1HF.cpp (hunk 2 above)
Comment (aloeliger) [reply L1Trigger/L1TCaloLayer1/test/testUCTLayer1HF.cpp]: Applied in most recent commit

**Comment by aloeliger** [comment]
PR will be good to go after retesting after applying the suggestions.

**Comment by aloeliger** [comment]
@mandrenguyen Do we need backports to 14_0?

**Comment by aloeliger** [comment]
please test

**Comment by aloeliger** [comment]
+l1
- I assume these RelVal inputs are not the fault of this PR as before

**Comment by mandrenguyen** [comment]
ignore tests-rejected with ib-failure

**Comment by mandrenguyen** [comment]
+1

**Pull request merged**